autodoc_traits.py

```diff
@@ -195,6 +187,20 @@ def get_doc(self):
         return super().get_doc()
 
 
+def hastraits_attrgetter(obj, name, *defargs):
+    """getattr for trait
+
+    Ensures when HasTraits are documented, their __doc__ attr is defined
+    as the .help string.
+    """
+    attr = safe_getattr(obj, name, *defargs)
+    if isinstance(attr, TraitType):
+        # ensure __doc__ is defined as the trait's help string
+        # if help is empty, that's the same as undocumented
+        attr.__doc__ = attr.help
```

Comment:
Having `__doc__` set here makes it pass filtering steps only if its a truthy string. In the end, whats set here is ignored as its determined by the TraitDocumenter's `get_doc` override instead.

Here are some options:

1. Remove TraitDocumenter's get_doc override and have it like this here, where the empty `.help` will make some traits be hidden by not being documented even if they can be configured.

   Outcome: Complexity is reduced, but we have compromised by no longer having a way to present configurable traitlets that aren't documented with a help string any more.

2. Remove TraitDocumenter's get_doc override and provide a default string saying that "No help string is provided." here.

   Outcome: Complexity is reduced, all configurable traits will show up still, and those without a .help string will have that notice that no help string is provided. The compromise is that we are forced to provide a default help string if provided with a falsy help string - which seems fine to me.

3. Let TraitDocumenter.get_doc remain, and set a dummy truthy string here no matter what, as setting it to `.help` when it will anyhow to be considered can be confusing.

I think option 2 is the best.

Reply:
A gif showing the impact of providing a truthy string or not.

![autodoc](https://user-images.githubusercontent.com/3837114/208171589-ae86738f-e315-4eec-b497-535e59649ce9.gif)

Reply:
Hmmm hmmm hmmm... Well, the downside is that we have controlled if non-configurable traits get documented like this before.

So we would maybe need something like...

```python
    if isinstance(attr, TraitType):
        # ensure __doc__ is defined as the trait's help string
        # if help is empty, that's the same as undocumented
        attr.__doc__ = attr.help or "No help string is provided."
        
        if not attr.metadata.get("config"):
            attr.__doc__ = ""
```

At the same time, if specific members is selected, or undoc-members is used, non-config traits with help strings will show up - without a help string even though they have one.

Reply:
I think the right call is to accept that non-config traits will show up if `:members:` is used.

Reply:
I went ahead and pushed a commit to be reviewed, we can discard it and thats fine as well!

In the commit f878007:

1. The `TraitDocumenter.get_doc` function override is removed. Going onward, the string set in `__doc__` must be truthy, otherwise will get filtered out. Also, if it is truty, it won't be filtered out - even if its a non-configurable trait.
2. The `hastraits_attrgetter` function is updated to provide a default value for the `.help` string if unset.

   ```python
           attr.__doc__ = attr.help or "No help string is provided."
   ```
3. The tests outcomes are updated. Non-configurable traits are now presented, with a help string or default help string, if `:members:` or `inherited-members:` options are passed.

---

I think as soon as we have removed the `get_doc` function, we have no way of letting non-configurable traits be ignored while also presenting their help strings if they are explicitly wished to be presented.

To not present non-configurable traits in the past, we have relied on two things:

1. That without `:members:` or `inherited-members:` options, they weren't added.
2. That without `__doc__`, they were filtered out.

Following EITHER https://github.com/ipython/traitlets/pull/816 or the change introduced in this commit, we will end up with non-configurable trait options showing up whenever `:members:` or `inherited-members:` is passed - as `__doc__` will be set.

I'm very happy with the behaviour we end up with now, and like this - we end up not being impacted by the https://github.com/ipython/traitlets/pull/816 change in a way that makes autodoc_traits behavde differently based on traitlets version.

Reply:
> I think the right call is to accept that non-config traits will show up if :members: is used.

I think this isn't a thing to 'accept', I think this is a bug that's fixed. Trait members behaving as if they were non-trait members is a goal, I think.

I'm happy with the behavior now, too! Thanks for updating the tests and further cleaning things up.

Reply:
Sorry, I'm not sure if I followed the conclusions. There's a lot going on. My understanding is that the goal is traits are handled consistently with attributes defined in other ways, to minimize complexity and confusion (traits _shouldn't_ be special, as much as possible):

- undocumented members are excluded by default
- documented members match handling of `:members:`

So the _only_ special handling left of traits is that _documented_ `config=True` traits are included by default when using `autoconfigurable`. If it's not documented _or_ it's not `config=True`, it's handled like every other attribute.

Reply:
I don't care much for how the behavior changes when using with `autoclass` as that is not something that is done anywhere. I'm :+1: for the change in autoclass to stop using a default truthy help string, but don't consider it so important.

I think its a notable change to stop presenting configurable traits without a truthy `help` value though, and that will be an outcome of removing the truthy default value for `autoconfigurable`. We could do that, but I would not consider it a pure bugfix. Suddenly undocumented traits that may be self-explanatory, for example booleans to toggle something on or off could very well not be provided a help string, and they would suddenly be hidden by `autoconfigurable`.

To conclude, I'm open to removing the truthy help string - but I think a major version change should be done for that. Since I've aimed to get a 1.1.0 out without needing to go for a breaking change, I'm opinionated towards getting this merged as it is for a 1.1.0 release, and then adding another dedicated PR focused on the breaking change part.

# Status summary

Normal attributes will not be affected by this PR no matter our choice below. They are ignored by `autoclass` and `autoconfigurable` unless `:undoc-members:` is used, then they are presented.

## `autoconfigurable`

What is required for a trait attribute to be presented when using `autoconfigurable`?

### Before PR and after PR in current state

Trait attribute | config | falsy config
-|-|-
**help** | always | `:members:`
**falsy help** | always | `:undoc-members:`

### After PR without default truthy help string

Trait attribute | config | falsy config
-|-|-
**help** | always | `:members:`
**falsy help** | `:undoc-members:` | `:undoc-members:`

## `autoclass`

What is required for a trait attribute to be presented when using `autoclass`?

### Before PR

Trait attribute | config | falsy config
-|-|-
**help** | `:undoc-members:` | `:undoc-members:`
**falsy help** | `:undoc-members:` | `:undoc-members:`

### After PR in current state

Trait attribute | config | falsy config
-|-|-
**help** | `:members:` | `:members:`
**falsy help** | `:members:` | `:members:`

### After PR without default truthy help string

Trait attribute | config | falsy config
-|-|-
**help** | `:members:` | `:members:`
**falsy help** | `:undoc-members:` | `:undoc-members:`

Reply:
Great point about the backward-compatibility! I think this is a good change in general, but I agree with keeping autoconfigurable behavior for a minor release. I'll take a stab at one more minor adjustment:

1. Only applying the include-undoc patch on autocnfigurable, but not all traits
2. Maybe a _deprecation_ warning for undocumented config being included by default, but no change in behavior.

That should keep behavior unbroken, paving the way for what I think will be the right behavior eventually. WDYT?

Reply:
Absolutely, sure!